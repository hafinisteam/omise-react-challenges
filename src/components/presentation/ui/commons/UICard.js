import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import COLOR from "~/utilities/layout/color";
import { mediaMax } from "~/utilities/layout";

const Card = styled.div`
  border-radius: 5px;
  box-shadow: -1px 5px 6px -2px #66728852;
  border: 1px solid #f2f2f2;
  overflow: hidden;
  position: relative;
`;


const CardMedia = styled.div`
  position: relative;
  height: 250px;
  overflow: hidden;
  img {
    width: 100%;
  }
  ${mediaMax.xl`
    height: 200px;
  `}
  ${mediaMax.md`
    height: 150px;
  `}
`;

CardMedia.displayName = "CardMedia";

export const CardBody = styled.div`
	position: relative;
	background: ${COLOR.white};
	padding: 20px;
	display: flex;
	align-items: center;
	justify-content: space-between;
	.card-title {
		font-size: 16px;
		font-weight: 600;
		color: ${COLOR.gray};
		margin-bottom: 0;
	}
	${mediaMax.xl`
    padding: 10px;
  `}
	${mediaMax.md`
    .card-title {
      font-size: 14px;
    }
  `}
	${mediaMax.sm`
    .card-title {
      font-size: 16px;
    }
  `}
`;

CardBody.displayName = "CardBody";

const UICard = props => {
  const { image, title, extraTitle, overlay } = props;
  return (
    <Card>
      {image && (
        <CardMedia>
          <img src={image} />
        </CardMedia>
      )}
      <CardBody>
        <h3 className="card-title">{title}</h3>
        {extraTitle && <div>{extraTitle}</div>}
      </CardBody>
      {overlay()}
    </Card>
  );
};

UICard.defaultProps = {
  image: null,
  title: "Card title",
  extraTitle: null,
  overlay: () => {}
};

UICard.propTypes = {
  image: PropTypes.string,
  title: PropTypes.string.isRequired,
  extraTitle: PropTypes.object
};

export default UICard;
