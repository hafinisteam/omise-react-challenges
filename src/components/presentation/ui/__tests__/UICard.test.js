// import { expect } from "chai";
import UICard from "../commons/UICard";
import "jest-styled-components";
import renderer from "react-test-renderer";

describe("<UICard />", () => {
  it("should render a CardBody with default props", () => {
    const wrapper = shallow(<UICard />);
    expect(wrapper.find("CardBody")).toHaveLength(1);
  });

  it("should render title correctly", () => {
    const wrapper = shallow(<UICard title="Test title" />);
    expect(wrapper.find(".card-title").text()).toEqual("Test title");
  });

  it("should render image when receive image props", () => {
    const wrapper = shallow(<UICard image="Test title" />);
    expect(wrapper.find("CardMedia")).toHaveLength(1);
  });

  it("render correctly with props", () => {
    const props = {
      title: "Ha Nguyen",
      image: "https://via.placeholder.com/350x150",
      extraTitle: <p>Extra</p>
    };
    const tree = renderer.create(<UICard {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
