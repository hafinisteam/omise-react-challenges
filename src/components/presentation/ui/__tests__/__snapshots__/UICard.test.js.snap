// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UICard /> render correctly with props 1`] = `
.c0 {
  border-radius: 5px;
  box-shadow: -1px 5px 6px -2px #66728852;
  border: 1px solid #f2f2f2;
  overflow: hidden;
  position: relative;
}

.c1 {
  position: relative;
  height: 250px;
  overflow: hidden;
}

.c1 img {
  width: 100%;
}

.c2 {
  position: relative;
  background: #fff;
  padding: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 .card-title {
  font-size: 16px;
  font-weight: 600;
  color: #667288;
  margin-bottom: 0;
}

@media (max-width:1200px) {
  .c1 {
    height: 200px;
  }
}

@media (max-width:768px) {
  .c1 {
    height: 150px;
  }
}

@media (max-width:1200px) {
  .c2 {
    padding: 10px;
  }
}

@media (max-width:768px) {
  .c2 .card-title {
    font-size: 14px;
  }
}

@media (max-width:576px) {
  .c2 .card-title {
    font-size: 16px;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <img
      src="https://via.placeholder.com/350x150"
    />
  </div>
  <div
    className="c2"
  >
    <h3
      className="card-title"
    >
      Ha Nguyen
    </h3>
    <div>
      <p>
        Extra
      </p>
    </div>
  </div>
</div>
`;
